src/std/misc/list.ss

```diff
@@ -268,23 +269,22 @@
     lst))
 
 ;; split the list lst into a list-of-lists using the unary procedure proc.
-;; (split '(1 2 "hi" 3 4) string?)                 => ((1 2) (3 4))
-;; (split '(1 2 a 3 4) (lambda (x) (equal? x 'a))) => ((1 2) (3 4))
-;; (split '(1 2 a 3 4) (cut equal? <> 'a))         => ((1 2) (3 4))
-;; (split [] number?)                              => ()
-(def (split lst proc)
-  (def (new-acc acc cur)
-    (if (pair? cur) (cons* (reverse! cur) acc) cur))
-  (let loop ((cur []) (acc []) (lst lst))
-    (cond
-     ((pair? lst)
-      (if (proc (car lst))
-	(loop [] (new-acc acc cur) (cdr lst))
-	(loop (cons (car lst) cur) acc (cdr lst))))
-     (else
-      (if (pair? cur)
-	(snoc (reverse! cur) acc)
-	acc)))))
+;; If limit is set, split the list only limit times.
+;;
+;; Examples:
+;;  (split '(1 2 "hi" 3 4) string?)                 => ((1 2) (3 4))
+;;  (split [1 2 0 3 4 0 5 6] zero? 1)               => ((1 2) (3 4 0 5 6))
+;;  (split [] number?)                              => ()
+(def (split lst proc (limit #f))
+  (if (pair? lst)
+    (let loop ((xs lst) (acc []) (buf []) (n (if (number? limit) limit -1)))
+      (match (cons (zero? n) xs)
```

Comment:
This is certainly convenient, but it also generates garbage on each iteration.
You can simply use `match*` and match both on the discriminant and the list without the additional consing.

So:
```
(match* ((zero? n) xs)
 ((#f [v . xs]) ...)
 ((#t [v . xs]) ...)
 (else ...)
```
```

Reply:
I tried to find something like that but didn't know `match*` existed, thanks.